src/webgpu/web_platform/canvas/readbackFromWebGPUCanvas.spec.ts

```diff
@@ -0,0 +1,265 @@
+
+    let snapshot: HTMLImageElement | ImageBitmap;
+    switch (t.params.snapshotType) {
+      case 'convertToBlob': {
+        if (typeof offscreenCanvas.convertToBlob === undefined) {
+          t.skip("Browser doesn't support OffscreenCanvas.convertToBlob");
+          return;
+        }
+        const blob = await offscreenCanvas.convertToBlob();
+        const url = URL.createObjectURL(blob);
+        const img = new Image(offscreenCanvas.width, offscreenCanvas.height);
+        img.src = url;
+        await raceWithRejectOnTimeout(img.decode(), 5000, 'load image timeout');
+        snapshot = img;
+        break;
+      }
+      case 'transferToImageBitmap': {
+        if (typeof offscreenCanvas.transferToImageBitmap === undefined) {
+          t.skip("Browser doesn't support OffscreenCanvas.transferToImageBitmap");
+          return;
+        }
+        snapshot = offscreenCanvas.transferToImageBitmap();
+        break;
+      }
+      case 'imageBitmap': {
+        snapshot = await createImageBitmap(offscreenCanvas);
+        break;
+      }
+      default:
+        unreachable();
+    }
+
+    await checkImageResult(t, snapshot, expect);
+  });
+
+g.test('canvas,uploadToWebGL')
+  .desc(
+    `
+    Ensure upload WebGPU context canvas to webgl texture is correct.
+    `
+  )
+  .params(u =>
+    u //
+      .combine('webgl', ['webgl', 'webgl2'])
+      .combine('upload', ['texImage2D', 'texSubImage2D'])
+  )
+  .fn(async t => {
+    const { webgl, upload } = t.params;
+    const canvas = (await initCanvasContent(t, 'onscreen')) as HTMLCanvasElement;
+
+    const expectCanvas: HTMLCanvasElement = createOnscreenCanvas(t, canvas.width, canvas.height);
+    const gl = expectCanvas.getContext(webgl) as WebGLRenderingContext | WebGL2RenderingContext;
+    if (gl === null) {
+      return;
+    }
+
+    const texture = gl.createTexture();
+    gl.bindTexture(gl.TEXTURE_2D, texture);
+    switch (upload) {
+      case 'texImage2D': {
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, canvas);
+        break;
+      }
+      case 'texSubImage2D': {
+        gl.texImage2D(
+          gl.TEXTURE_2D,
+          0,
+          gl.RGBA,
+          canvas.width,
+          canvas.height,
+          0,
+          gl.RGBA,
+          gl.UNSIGNED_BYTE,
+          null
+        );
+        gl.texSubImage2D(gl.TEXTURE_2D, 0, 0, 0, gl.RGBA, gl.UNSIGNED_BYTE, canvas);
+        break;
+      }
+      default:
+        unreachable();
+    }
+
+    const fb = gl.createFramebuffer();
+
+    gl.bindFramebuffer(gl.FRAMEBUFFER, fb);
+    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, texture, 0);
+
+    const pixels = new Uint8Array(canvas.width * canvas.height * 4);
+    gl.readPixels(0, 0, 2, 2, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
+    const actual = new Uint8ClampedArray(pixels);
+
+    t.expectOK(checkElementsEqual(actual, webglExpect));
+  });
```

Comment:
We should have one more test for using a webgpu canvas as a source to 2d canvas `drawImage`

Reply:
Ah, missing this. Added, thanks!